lib/auth/touchid/api.go

```diff
@@ -0,0 +1,376 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	credentialID := resp.CredentialID
+	pubKeyRaw := resp.publicKeyRaw
+
+	// Parse public key and transform to the required CBOR object.
+	pubKey, err := pubKeyFromRawAppleKey(pubKeyRaw)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	x := make([]byte, 32) // x and y must have exactly 32 bytes in EC2PublicKeyData.
+	y := make([]byte, 32)
+	pubKey.X.FillBytes(x)
+	pubKey.Y.FillBytes(y)
+
+	pubKeyCBOR, err := cbor.Marshal(
+		&webauthncose.EC2PublicKeyData{
+			PublicKeyData: webauthncose.PublicKeyData{
+				KeyType:   int64(webauthncose.EllipticKey),
+				Algorithm: int64(webauthncose.AlgES256),
+			},
+			// See https://datatracker.ietf.org/doc/html/rfc8152#section-13.1.
+			Curve:  1, // P-256
+			XCoord: x,
+			YCoord: y,
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	attData, err := makeAttestationData(
+		protocol.CreateCeremony, origin, rpID, cc.Response.Challenge,
+		&credentialData{
+			id:         credentialID,
+			pubKeyCBOR: pubKeyCBOR,
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sig, err := native.Authenticate(credentialID, attData.digest)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	attObj, err := cbor.Marshal(protocol.AttestationObject{
+		RawAuthData: attData.rawAuthData,
+		Format:      "packed",
+		AttStatement: map[string]interface{}{
+			"alg": int64(webauthncose.AlgES256),
+			"sig": sig,
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &wanlib.CredentialCreationResponse{
+		PublicKeyCredential: wanlib.PublicKeyCredential{
+			Credential: wanlib.Credential{
+				ID:   credentialID,
+				Type: string(protocol.PublicKeyCredentialType),
+			},
+			RawID: []byte(credentialID),
+		},
+		AttestationResponse: wanlib.AuthenticatorAttestationResponse{
+			AuthenticatorResponse: wanlib.AuthenticatorResponse{
+				ClientDataJSON: attData.ccdJSON,
+			},
+			AttestationObject: attObj,
+		},
+	}, nil
+}
+
+func pubKeyFromRawAppleKey(pubKeyRaw []byte) (*ecdsa.PublicKey, error) {
+	// Verify key length to avoid a potential panic below.
+	// 3 is the smallest number that clears it, but in practice 65 is the more
+	// common length.
+	// Apple's docs make no guarantees, hence no assumptions are made here.
+	if len(pubKeyRaw) < 3 {
+		return nil, fmt.Errorf("public key representation too small (%v bytes)", len(pubKeyRaw))
+	}
+
+	// "For an elliptic curve public key, the format follows the ANSI X9.63
+	// standard using a byte string of 04 || X || Y. (...) All of these
+	// representations use constant size integers, including leading zeros as
+	// needed."
+	// https://developer.apple.com/documentation/security/1643698-seckeycopyexternalrepresentation?language=objc
+	pubKeyRaw = pubKeyRaw[1:] // skip 0x04
+	l := len(pubKeyRaw) / 2
+	x := pubKeyRaw[:l]
+	y := pubKeyRaw[l:]
+
+	return &ecdsa.PublicKey{
+		Curve: elliptic.P256(),
+		X:     (&big.Int{}).SetBytes(x),
+		Y:     (&big.Int{}).SetBytes(y),
```

Comment:
Do we want to be extra safe and check `elliptic.P256().IsOnCurve(X, Y)` here? It comes from the user's secure enclave, so presumably it's already good.

Reply:
Yeah, not much of a point double-checking the Enclave, I think.